vault/hcp_link/link.go

```diff
@@ -0,0 +1,375 @@
+	invalidClientCredentials    = errors.New("failed to get access token: oauth2: cannot fetch token: 401 Unauthorized")
+)
+
+type HCPLinkVault struct {
+	l            sync.Mutex
+	LinkStatus   internal.WrappedCoreHCPLinkStatus
+	scadaConfig  *scada.Config
+	linkConfig   *linkConfig.Config
+	link         link.Link
+	logger       hclog.Logger
+	capabilities map[string]capabilities.Capability
+	stopCh       chan struct{}
+	running      bool
+}
+
+var _ HCPLinkVaultInterface = &HCPLinkVault{}
+
+func NewHCPLink(linkConf *configutil.HCPLinkConfig, core *vault.Core, logger hclog.Logger) (*WrappedHCPLinkVault, error) {
+	if linkConf == nil {
+		return nil, nil
+	}
+
+	scadaLogger := logger.Named("scada")
+	scadaConfig, err := internal.NewScadaConfig(linkConf, scadaLogger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to instantiate SCADA config, %w", err)
+	}
+
+	// setting the link status in core, as link config is not nil
+	// At this point scada provider has not been started yet
+	// After starting scada provider, we need to use
+	// scadaProvider.SessionStatus() to get the status of the connection
+	core.SetHCPLinkStatus(
+		buildConnectionErrorMessage(scada.SessionStatusDisconnected, scada.ErrProviderNotStarted.Error(), time.Now()),
+		scadaConfig.Resource.Location.ProjectID,
+	)
+
+	// Creating SCADA provider
+	scadaProvider, err := scada.New(scadaConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to instantiate SCADA provider: %w", err)
+	}
+
+	resource := scadaConfig.Resource
+	hcpConfig := scadaConfig.HCPConfig
+	version := vaultVersion.Version
+
+	// initializing node status reporter. This capability is configured by link lib.
+	statusReporter := &node_status.NodeStatusReporter{
+		NodeStatusGetter: core,
+	}
+	nodeID, err := core.LoadNodeID()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get nodeID, %w", err)
+	}
+
+	// Compile the Link config
+	var conf *linkConfig.Config
+	conf, err = internal.NewLinkConfig(
+		nodeID,
+		version,
+		resource,
+		scadaProvider,
+		hcpConfig,
+		statusReporter,
+		logger,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to instantiate Link library config: %w", err)
+	}
+
+	// Create a Link library instance
+	hcpLink, err := link.New(conf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to instantiate Link library: %w", err)
+	}
+
+	hcpLinkCaps, err := initializeCapabilities(linkConf, scadaConfig, scadaProvider, core, logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize capabilities: %w", err)
+	}
+
+	hcpLinkVault := &HCPLinkVault{
+		LinkStatus:   core,
+		scadaConfig:  scadaConfig,
+		linkConfig:   conf,
+		link:         hcpLink,
+		capabilities: hcpLinkCaps,
+		stopCh:       make(chan struct{}),
+		logger:       logger,
+	}
+
+	// Start hcpLink and ScadaProvider
+	err = hcpLinkVault.start()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start hcp link, %w", err)
+	}
+
+	wrapped := &WrappedHCPLinkVault{hcpLinkVault}
```

Comment:
I wonder if we still need this wrapped struct. IIRC, we used it to be able to introduce a stub for OSS. Now, that these are ported over, do we still need this?

Reply:
That's a really good point. I don't think so.